Upper tx validity could be very big on close/contest

v0d1ch opened the pull request:
## Why

If the tx has too high upper validity bound it can cause large wait times since we are using this upper validity when checking the expected deadline. 

## What 

Take a look at the [ADR21](https://github.com/input-output-hk/hydra/blob/aafa0592b68a26c4f28bfee0b0481fb4022b528c/docs/adr/2022-12-05_021-Bounded-transaction-validity-on-Hydra-protocol-transactions.md) for context

![Screenshot 2022-12-08 14:25:58](https://user-images.githubusercontent.com/1666357/206459205-0f04f5a0-e5ee-4474-b788-05fd144ef9f2.png)



To check before merging:
* [x] CHANGELOG is up to date
* [x] Up to date with master

ch1bo committed:
Create a mutation for close upper bound being far in the future

v0d1ch committed:
Start with the script changes

ch1bo committed:
Add the check for tMax/tMin in the close validator

v0d1ch committed:
Introduce lower tx bound for close txs

v0d1ch committed:
Make the _is healthy_ test pass

v0d1ch committed:
Use arbitrary for contestation period in Close tests

ffakenz committed:
Minor fix to txcost, add missing slotno argg

ffakenz committed:
Add missing changes from master rebase

v0d1ch committed:
Generate SnapshotNumber also

v0d1ch committed:
Some renaming

v0d1ch committed:
Add breaking change to changelog

v0d1ch committed:
Cleanup

v0d1ch committed:
Specify lower/upper bounds for simple tx in tests

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -159,6 +159,7 @@ mkSimpleTx (txin, TxOut owner valueIn datum refScript) (recipient, valueOut) sk
       { txIns = map (,BuildTxWith $ KeyWitness KeyWitnessForSpending) [txin]
       , txOuts = outs
       , txFee = TxFeeExplicit fee
+      , txValidityRange = (TxValidityLowerBound 0, TxValidityUpperBound 1)
```
Why are we doing this? It will make transactions created by `mkSimpleTx` only validate on slot 0 and 1 (not sure if these are inclusive).

v0d1ch replied:
We needed to provide validity ranges so the validator works. I tried using the network tip here but it seems that in this scenario onchain part always sees this tx to be valid starting at slot 0.

v0d1ch replied:
This tx is actually not used in the on-chain validation so I'll revert this change.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -386,9 +386,10 @@ close ::
   OpenState ->
   ConfirmedSnapshot Tx ->
   PointInTime ->
+  SlotNo ->
```
Document, swap sequence and add a note about how much apart these two arguments (slots) can be.

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -38,6 +38,9 @@ instance (Arbitrary tx, Arbitrary (UTxOType tx)) => Arbitrary (Snapshot tx) wher
     ]
 
+genSnapShot :: Gen SnapshotNumber
+genSnapShot = arbitrary
```
This is weird, but acceptable if needed. Maybe we can get rid of overlapping instances by not importing the `Natural` orphans.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -87,8 +92,8 @@ healthyCloseUTxO =
-healthySnapshotNumber :: SnapshotNumber
-healthySnapshotNumber = 1
+genSnapshotNumber :: SnapshotNumber
+genSnapshotNumber = genSnapShot `generateWith` 42
```
Beware: this does not generate arbitrary values on different test runs. It's basically just picking some random number, which might change whenever we change the code, but not multiple values within one version of the code.

v0d1ch replied:
I've reverted this in the end anyway.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -99,10 +104,12 @@ healthyCloseDatum =
+-- NB: We don't want to wait too long for contestation period to pass
+-- so constraining this to < 100
+genContestationPeriodSeconds :: Integer
+genContestationPeriodSeconds = arbitrary `suchThat` (< 100) `generateWith` 42
```
These `gen...` functions are not really generators. Rename them back to indicate that they are just picked values (see comment above)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -273,7 +273,8 @@ fromPostChainTx timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
     (CloseTx{confirmedSnapshot}, Open st) -> do
       shifted <- throwLeft $ adjustPointInTime closeGraceTime pointInTime
-      pure (close ctx st confirmedSnapshot shifted)
+      -- REVIEW: what do we use for lower bound slot here?
```
The current slot (maybe name it like this) should be fine.

ch1bo replied:
But maybe.. it's not. It may be that the ledger is still not at the `currentPointInTime` and the node might reject this slot.

ch1bo replied:
Option 1: Change `queryTimeHandle` to use the latest slot known to the node as `currentPointInTime`. BUT we need to check all usages of time handle whether this change in semantics is fine!

Option 2: Use the upper bound (`shifted` here) and adjust get a slot number within the allowed window (= contestation period), i.e. something like `startSlotNo = upperSlotNo - contestionPeriodInSlots`. NOTE we need access to the right contestation period for this option.

v0d1ch replied:
Note to self: I didn't explore but want to the Option no 1 here

ch1bo committed:
Detail haddock on queryTimeHandle

ch1bo committed:
Make arbitrary ContstationPeriod more demanding

We did not really yield 0s and 1s values before. Although they are
important boundary values for contestation period. Especially as we are
tying it up with the lower/upper validity bounds of the close
transaction now.

v0d1ch committed:
Calculate start slot by using contestation period

v0d1ch committed:
Reorder the parameters for close and add docs

v0d1ch committed:
Remove tx validity from simple off-chain tx

v0d1ch committed:
Bring back the old functions' names in Close

v0d1ch committed:
Rearange the parameters for closeTx

v0d1ch committed:
Checkpoint

v0d1ch committed:
Make relevant test pass

v0d1ch committed:
Fix is healthy close test

v0d1ch committed:
Redundant import

v0d1ch committed:
Calculate tx start bound from end bound and offchain cp

v0d1ch committed:
Add adr 21

ch1bo committed:
Ensemble clarification on ADR21

v0d1ch committed:
Get adr21 in shape for reviewing

v0d1ch committed:
Rename the adr21

ch1bo commented on docs/adr/CloseTx-validity-bounds-and-deadline-calculation-DoS.md:
```diff
@@ -0,0 +1,73 @@
+---
+slug: 21
```
Make sure to follow the naming scheme of other adr documents.

ch1bo commented on docs/adr/CloseTx-validity-bounds-and-deadline-calculation-DoS.md:
```diff
@@ -0,0 +1,73 @@
+* Need to check the contestation period of an `InitTx` and warn the user if it does not match.
+  - Ideally we would error out not issue a warning but: This is super hard in the current code!
+    The chain layer can only decide to observe or not (`Maybe`) and informing the user / taking action on this is messy.
+* Make the `contestationPeriod` in the `Init` client input not configurable via the API.
```
I do agree with the contents of this, but it feels that some things from the context are actually decisions and some decision bullets read a lot like consequences (and are duplicate). For example, the fact that we set `T_min` on `closeTx` is a decision we take - although motivated by wanting correct scripts. The way we do set it is another decision, but the fact that the contestation period now needs to be meaningful for the network we are on is a consequence.

One decision missing (maybe having consequences?): Remove the `closeGraceTime`

v0d1ch committed:
Rename the adr21 file

v0d1ch committed:
Introduce new hydra-node flag

pgrange committed:
FIX tests

ffakenz committed:
Add ContestationPeriod to Environment and usit during onIdleClientInit, ignoring the one from the Init tx

v0d1ch committed:
Use newtype for SnapshotNumber

pgrange committed:
Exploring on how to make the tests pass

v0d1ch committed:
Rename fromPostChainTx to prepareTxToPost

v0d1ch committed:
Re-use getTimeHandle and stub out the contestation-period flag

v0d1ch committed:
Fix RunOptions json

v0d1ch committed:
Fix roundtrip parsing for RunOptions

v0d1ch committed:
Have a default value for contestation period flag

pgrange committed:
Contestation Period is now a cli parameter

The tests need to setup the contestation period when instanciating
a new Hydra node and not only when calling the API.

v0d1ch committed:
Use current slot ad tx min bound

pgrange committed:
FIX Warnings

pgrange committed:
Make the contestation period part of the chain context

So that we can use it later when building close transactions for
:x

pgrange committed:
FIX un-commited changes

v0d1ch committed:
Fix all warnings

v0d1ch committed:
Remove closeGraceTime and instead use contestationPeriod

v0d1ch committed:
Remove contestationPeriod from Init client input

pgrange committed:
FIX Options tests

v0d1ch committed:
Remove contestation period from api.yaml

v0d1ch committed:
Fix a typo and improve on argument naming

v0d1ch committed:
Thread the contestationPeriod correctly in all of the tests

v0d1ch committed:
Make sure to use positive number for starting slot in genCloseTx

v0d1ch committed:
Reduce the wait time to make the tests pass

v0d1ch committed:
Ignore InitTx with the _wrong_ contestation period

v0d1ch committed:
Update adr21 wording to exclude the stuff we are not doing in this PR

v0d1ch committed:
Bring back the old name for genContestationDeadline

v0d1ch committed:
Always use contestation period from underlying context

v0d1ch committed:
Remove un-needed function

v0d1ch committed:
Seed the contestation period in MBT

v0d1ch committed:
Set adr21 status to accepted

v0d1ch committed:
Update the changelog

v0d1ch committed:
Add authors to docusaurus

ffakenz commented on docs/adr/authors.yml:
```diff
@@ -0,0 +1,8 @@
+  "Franco Testagrossa":{ "imageURL": "https://avatars.githubusercontent.com/u/14299604?v=4"},
+  "Pascal Grange": { "imageURL": "https://avatars.githubusercontent.com/u/378506?v=4"},
+  "Sasha Bogicevic":{"imageURL": "https://avatars.githubusercontent.com/u/1666357?v=4"}
+}
```
❤️

ffakenz commented on hydra-plutus/hydra-plutus.cabal:
```diff
@@ -137,7 +137,7 @@ test-suite tests
     , QuickCheck
     , time
 
-  build-tool-depends: hspec-discover:hspec-discover -any
```
what is the benefit we gain by removing this `-any` flag?

v0d1ch replied:
I guess then at compilation time cabal will pick exact version of hspec-discover it has in scope. This is something we did in ensamble I think, can't really remember but I do prefer that we don't pick any random version that works of hspec-discover any way.

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -47,6 +47,9 @@ import qualified PlutusTx.AssocMap as Map
 
 -- REVIEW: Functions not re-exported "as V2", but using the same data types.
+
+-- REVIEW: Functions not re-exported "as V2", but using the same data types.
```
maybe remove, this is the same as line #49

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -271,8 +276,12 @@ checkClose ::
+    && checkSnapshot
+    && mustBeSignedByParticipant ctx headContext
  where
+  hasBoundedValidity = traceIfFalse "hasBoundedValidity check failed" $ tMax - tMin <= cp
```
is it possible for `tMin` to be greater than `tMax`, or that is checked somewhere else?

v0d1ch replied:
You can't post a tx that doesn't have the correct validity I think. You could omit the lower/upper bound but if you set it I would assume that it would get rejected by the node. We can write a test to prove it!

v0d1ch replied:
added a mutation test for this case also

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -294,9 +303,20 @@ checkClose ctx headContext parties initialUtxoHash snapshotNumber closedUtxoHash
             && checkHeadOutputDatum ctx expectedOutputDatum
     | otherwise = traceError "negative snapshot number"
+
+  cp = fromMilliSeconds (milliseconds cperiod)
```
can cp be zero?

v0d1ch replied:
It can if you set it to be zero. Another thought - write a test where cp is 0 .

ffakenz replied:
0 is member of Natural numbers, but i am not sure we should accept a cp = 0

v0d1ch replied:
I added a mutation test that exercises zero value for CP

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -95,22 +96,22 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
   refuelIfNeeded tracer node Alice 100_000_000
   -- Start hydra-node on chain tip
   tip <- queryTip networkId nodeSocket
+  let contestationPeriod = UnsafeContestationPeriod 2
```
is there a reason why this is 2 and not 1 as the rest?

maybe like in TUI we can create a commented defaultContestationPeriod to re-use.

pgrange replied:
Good catch. The idea of this P.R. is to keep the tests as they were before. So we should not change 1 (old like 103) by 2 here.

```suggestion
  let contestationPeriod = UnsafeContestationPeriod 1
```

v0d1ch replied:
This is a left-over, I'll bring it back to 1. The changes that actually made this test work are a bit below where I reduced the amount of time we wait from 600 to 10.

ffakenz commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -122,7 +123,8 @@ spec = around showLogsOnFailure $ do
 
               hydraScriptsTxId <- publishHydraScriptsAs node Faucet
-              withHydraCluster tracer tmpDir nodeSocket firstNodeId cardanoKeys hydraKeys hydraScriptsTxId $ \nodes -> do
+              let contestationPeriod = UnsafeContestationPeriod 2
```
is there a reason why this is 2 and not 1 as the rest?

maybe like in TUI we can create a commented defaultContestationPeriod to re-use.

v0d1ch replied:
We are using 2 because this test was previously using the same value https://github.com/input-output-hk/hydra/pull/615/files/d91550eeb74deee5c4ee337cd3942872ffa3d39d#diff-f3f9817ebc889ef314514ecb08ab7ffec2d5454063e34d381c2b1c8b0daca830L134

ffakenz commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -623,6 +623,11 @@ partyKeys =
 
+genContestationPeriod :: Gen ContestationPeriod
+genContestationPeriod = do
+  i :: Word <- arbitrary `suchThat` (> 0)
```
i think i will also limit the upper bound.

v0d1ch replied:
What value do you think it would make sense for upper bound?

ffakenz replied:
idk, i just dont want our test to run forever :P

ffakenz commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -97,6 +97,7 @@ data GlobalState
   | Idle
       { idleParties :: [Party]
       , cardanoKeys :: [VerificationKey PaymentKey]
+      , idleContestationPeriod :: ContestationPeriod
```
why?
is this part of the `WorldState`?

v0d1ch replied:
It made sense since now we configure our hydra-node on start and rely on this value in the rest of the code. `Idle` is similar in a sense you need to provide values like parties, keys etc. before trying to run the tests.

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -559,15 +562,18 @@ data InitObservation = InitObservation
   (ix, headOut, headData, Head.Initial cp ps) <- findFirst headOutput indexedOutputs
   parties <- mapM partyFromChain ps
   let contestationPeriod = fromChain cp
+  guard $ ourNode'sCP == contestationPeriod
```
💂

ffakenz commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -263,6 +264,17 @@ genPointInTime = do
+-- is not higher than the cp
+genSlotWithPointInTimeFromCP :: Word64 -> Gen (SlotNo, (SlotNo, UTCTime))
+genSlotWithPointInTimeFromCP contestationPeriod = do
+  startSlot@(SlotNo start) <- SlotNo <$> (arbitrary `suchThat` (> 0))
```
shall we define an upper bound here?

v0d1ch committed:
Address minor PR review comments

v0d1ch committed:
Use choose instead of suchThat

v0d1ch committed:
Add mutation where tx lower bound is higher than the upper bound

v0d1ch committed:
Add mutation where CP is zero

v0d1ch committed:
Change tx upper validity calculation

ch1bo commented on docs/adr/authors.yml:
```diff
@@ -0,0 +1,8 @@
+{
+  "Sebastian Nagel": { "imageURL": "https://avatars.githubusercontent.com/u/2621189?v=4"},
```
We have two authors files now. We should re-use the `docs/authors.yaml` and extend it accordingly.

ch1bo commented on CHANGELOG.md:
```diff
@@ -26,6 +26,13 @@ changes.
 
+- **BREAKING** Implemented [ADR21](https://hydra.family/head-protocol/adr/21) to introduce bounded close transactions:
+  + There is a new hydra-node flag `--contestation-period` to control the close tx validity bounds as well as
+    determine the contestation deadline.
```
Can you explain what the node does to users? Before they could provide the contestation period on the `Init` input. Now this is different. The deadline is always `2 * --contestation-period` after the `Close`.

ch1bo commented on CHANGELOG.md:
```diff
@@ -26,6 +26,13 @@ changes.
+- **BREAKING** Implemented [ADR21](https://hydra.family/head-protocol/adr/21) to introduce bounded close transactions:
+  + There is a new hydra-node flag `--contestation-period` to control the close tx validity bounds as well as
+    determine the contestation deadline.
+  + If hydra-node receives a `Init` tx with the _wrong_ contestation period then this tx is ignored which implies
```
```suggestion
  + If hydra-node receives a `Init` tx with a contestation period _not matching_ `--contestation-period`, this tx is ignored which implies
```

ch1bo commented on CHANGELOG.md:
```diff
@@ -26,6 +26,13 @@ changes.
   + Changed logs of `BeginInitialize` and `EndInitialize`.
   + Added `SkipUpdate` constructor to the wallet logs.
 
+- **BREAKING** Implemented [ADR21](https://hydra.family/head-protocol/adr/21) to introduce bounded close transactions:
```
The fact that this is ADR21 is a bit of a side information. It's much more important to users that we essentially changed the way the contestation deadline is computed (and added a new command line option).

ch1bo commented on hydra-node/src/Hydra/ContestationPeriod.hs:
```diff
@@ -5,12 +5,12 @@ import Hydra.Prelude hiding (Show, show)
 -- | A positive, non-zero number of seconds.
 newtype ContestationPeriod = UnsafeContestationPeriod Natural
-  deriving (Eq)
+  deriving (Eq, Read)
```
We don't use `Read`, especially not if it's inconsistent with `Show`!

Should: Remove `Read` instance and parse it "by hand"

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +556,27 @@ hydraNodeCommand =
+        <> showDefault
+        <> completer (listCompleter ["1", "2", "42"])
+        <> help
+          "Contestation period for close transaction. \
```
+ "in seconds".

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +556,27 @@ hydraNodeCommand =
+        <> metavar "CONTESTATION-PERIOD"
+        <> value defaultContestationPeriod'
+        <> showDefault
+        <> completer (listCompleter ["1", "2", "42"])
```
`1` and `2` are both not very sensible values. Neither is `3` as the default `value`. Please pick something meaningful here. Maybe `60` seconds (at minimum)?

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +556,27 @@ hydraNodeCommand =
+defaultContestationPeriod' :: Natural
+defaultContestationPeriod' = 3
+
+contestationPeriodParser :: Parser Natural
```
```suggestion
contestationPeriodParser :: Parser ContestationPeriod
```

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +556,27 @@ hydraNodeCommand =
+contestationPeriodParser :: Parser Natural
+contestationPeriodParser =
+  option
+    auto
```
Besides simple numeric values, this should also parse `show`n values of `ContestationPeriod`. Currently this is for example: `3s`.

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -650,6 +677,7 @@ toArgs
       ["--network-id", toArgNetworkId networkId]
         <> ["--node-socket", nodeSocket]
         <> ["--cardano-signing-key", cardanoSigningKey]
+        <> ["--contestation-period", show cp]
```
If you properly implement the `--contestatio-period` parser to also parse `show`n values, this can be just `show contestationPeriod` and does not require the usage of `UnsafeContestationPeriod` (it's called `Unsafe` for a reason :)

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -263,6 +264,17 @@ genPointInTime = do
+-- | Parameter here is the contestation period (cp) so we need to generate
+-- start (tMin) and end (tMax) tx validity bound such that their difference
+-- is not higher than the cp
+genSlotWithPointInTimeFromCP :: Word64 -> Gen (SlotNo, (SlotNo, UTCTime))
```
Should: use the type `ContestationPeriod`

Also, please indicate that the returned slots are validity bounds. So maybe calling this `genValidityBoundsFromContestationPeriod`?

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -263,6 +264,17 @@ genPointInTime = do
+-- is not higher than the cp
+genSlotWithPointInTimeFromCP :: Word64 -> Gen (SlotNo, (SlotNo, UTCTime))
+genSlotWithPointInTimeFromCP contestationPeriod = do
+  startSlot@(SlotNo start) <- SlotNo <$> choose (0, 100)
```
Why not `arbitrary` slot numbers?

v0d1ch replied:
@ffakenz had a comment about constraining these values so we don't wait too long  for close and fanout txs in tests. It also works with arbitrary values so leaving it as arbitrary for now until we detect long running tests.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -834,18 +850,22 @@ genCloseTx numParties = do
   snapshot <- genConfirmedSnapshot 0 u0 (ctxHydraSigningKeys ctx)
-  pointInTime <- genPointInTime
+  let (UnsafeContestationPeriod cp) = ctxContestationPeriod ctx
+      contestationPeriodSlotNo = fromIntegral cp
```
This is an *unsafe* conversion of seconds to slots!! It's even wrong for our devnet configuration!

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -834,18 +850,22 @@ genCloseTx numParties = do
   cctx <- pickChainContext ctx
-  pure (cctx, stOpen, close cctx stOpen snapshot pointInTime, snapshot)
+  (startSlot, pointInTime) <- genSlotWithPointInTimeFromCP contestationPeriodSlotNo
+  -- XXX: The difference between startSlot and pointInTime needs to be <= contestationPeriod
```
Why the `XXX`?

v0d1ch replied:
Should have been just the comment.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -834,18 +850,22 @@ genCloseTx numParties = do
   cctx <- pickChainContext ctx
-  let txClose = close cctx stOpen confirmed closePointInTime
+  let (UnsafeContestationPeriod cp) = Hydra.Chain.Direct.State.contestationPeriod cctx
+  (startSlot, closePointInTime) <- genSlotWithPointInTimeFromCP $ fromIntegral cp
```
This also interprets contestation period seconds as slots and is wrong for our devnet!

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -897,9 +917,10 @@ genStClosed ctx utxo = do
   cctx <- pickChainContext ctx
-  let txClose = close cctx stOpen snapshot pointInTime
+  let (UnsafeContestationPeriod cp) = Hydra.Chain.Direct.State.contestationPeriod cctx
+  (startSlot, pointInTime) <- genSlotWithPointInTimeFromCP (fromIntegral cp)
```
Same here

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -559,15 +562,18 @@ data InitObservation = InitObservation
   Tx ->
   Maybe InitObservation
-observeInitTx networkId cardanoKeys party tx = do
+observeInitTx networkId cardanoKeys ourNode'sCP party tx = do
```
This is not a good variable name (IMO). How about `expectedCP`?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -782,7 +788,7 @@ observeCloseTx utxo tx = do
                 }
           , headId
-          , snapshotNumber
+          , snapshotNumber = UnsafeSnapshotNumber sn
```
Defining a function `SnapshotNumber.fromChain` or so could make it clear that this is a fine conversion?

v0d1ch replied:
I've named it `fromChainSnapshot` since we have `fromChain` already in `ContestationPeriod`

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -56,21 +56,22 @@ import Hydra.Ledger (IsTx, Ledger)
       { party = deriveParty sk
       , signingKey = sk
       , otherParties
+      , contestationPeriod = contestationPeriod
```
```suggestion
      , contestationPeriod
```

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,19 @@ chainSyncHandler tracer callback getTimeHandle ctx =
-fromPostChainTx timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
-  pointInTime <- throwLeft currentPointInTime
+prepareTxToPost timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
+  pointInTime@(currentSlot, _) <- throwLeft currentPointInTime
```
This will fail on a real network! Using the `currentSlot` (as it is currently computed by `queryTimeHandle` - this is where `currentPointInTime` is coming from) will be rejected by a cardano-node that has not seen a block with a slot `>` than this one in it!

Must: Change `queryTimeHandle` to actually use the last known slot (== tip) as reported by the node as `currentSlot`. Also, check whether we rely on the `currentSlot` being the projected, wall clock slot -> we would break this code when we change the semantics of `queryTimeHandle`.

ch1bo reviewed (changes requested):
This will not work as is.. have you tried running this against a non-devnet? See my comment on the `currentSlot`. This and the assumed conversion from contestation period (seconds) to slots needs to be fixed.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -623,6 +623,11 @@ partyKeys =
 
+genContestationPeriod :: Gen ContestationPeriod
+genContestationPeriod = do
+  i :: Word <- choose (0, 100)
```
Why not arbitrary?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -364,8 +363,9 @@ instance
         performNewTx party transaction
-      Init party contestationPeriod ->
-        party `sendsInput` Input.Init{contestationPeriod}
+      -- TODO: seems that contestationPeriod is redundant here?
```
```suggestion
```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -170,11 +177,17 @@ genCloseMutation (tx, _utxo) =
     , SomeMutation MutateCloseUTxOHash . ChangeOutput 0 <$> mutateCloseUTxOHash
-    , SomeMutation MutateCloseContestationDeadline . ChangeOutput 0 <$> mutateClosedContestationDeadline
+    , SomeMutation MutateCloseContestationDeadline . ChangeOutput 0 <$> mutateClosedContestationDeadline Nothing
+    , SomeMutation MutateCloseContestationDeadline . ChangeOutput 0 <$> mutateClosedContestationDeadline (Just 0)
```
Do not reuse the label. These are two different mutations / attack vectors, right?

v0d1ch replied:
Not really different. I added parameter of type `Maybe Integer` so if you that you can control which value gets set as contestation deadline/or just pick arbitrary.

ch1bo replied:
But do we have two different generators for this? Each `SomeMutation` is a generator and if both cases interest you (and are cases of `MutateCloseContestationDeadline`), and seemingly *any one of them* is enough.. then a simple generator is what you want here.

If you care about both being tasted -> two cases / two labels.

v0d1ch committed:
Fix Model changes after rebase

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,19 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+  pointInTime@(currentSlot, _) <- throwLeft currentPointInTime
+  let (UnsafeContestationPeriod cpNatural) = contestationPeriod ctx
+      -- we are using the contestationPeriod as tx close grace time. See ADR21 for context
+      closeGraceTime = SlotNo $ fromIntegral $ cpNatural * 2
```
This should not be 2*cp! The validator should fail on anything `> cp`

v0d1ch replied:
Yes, with the new changes this is no longer valid/needed.

v0d1ch replied:
BUT we need to bump contestation period in end-to-end tests from value 1 to 2 for example since I observed failed tests where upper tx bound ends up being the same slot in which the validation happens.

v0d1ch committed:
Use current tip instead of wall-clock time

v0d1ch committed:
Parse contestation-period argument and introduce a test case

v0d1ch committed:
Revisit contestation-period parsing

v0d1ch committed:
Generate positive values for contestation period in RunOptions

v0d1ch committed:
Use arbitrary for cp in MBT

v0d1ch committed:
Fix the failing MBT test

v0d1ch committed:
Add more authors and reuse the same file

v0d1ch committed:
Rename genSlotWithPointInTimeFromCP to genValidityBoundsFromContestationPeriod

v0d1ch committed:
Thread ContestationPeriod instead of Word value for genValidityBoundsFromContestationPeriod

v0d1ch committed:
Use arbitrary values for SlotNo

v0d1ch committed:
Fix for call to genValidityBoundsFromContestationPeriod

v0d1ch committed:
Modify changelog

v0d1ch committed:
No need to multiply cp by 2

v0d1ch committed:
Introduce fromChainSnapshot function

v0d1ch committed:
Fix for e2e tests. CP no longer affects just the deadline but also tx validity

v0d1ch committed:
Use throwIO instead of error

v0d1ch committed:
Ch-ch-ch-change changelog

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -42,33 +44,32 @@ data TimeHandle = TimeHandle
   uptimeSeconds <- getPositive <$> arbitrary
-  let uptime = secondsToNominalDiffTime uptimeSeconds
-      currentTime = addUTCTime uptime startTime
+  let currentSlotNo = SlotNo $ truncate $ uptimeSeconds + startSeconds
```
I am having doubts that this is correct implementation? Tests are green but seems incorrect to just assign the current slot to whatever start + uptime seconds are? cc @ch1bo

v0d1ch replied:
But horizonSlot is constructed in the similar manner so I guess it can be actually correct.

ch1bo replied:
Yeah.. this is only okay to do here as we have a fixed slot length of 1 second. See haddock of `eraHistoryWithHorizonAt`. Maybe worthwhile to comment this here.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,19 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+  pointInTime@(currentSlot, _) <- throwLeft currentPointInTime
+  let (UnsafeContestationPeriod cpNatural) = contestationPeriod ctx
+      -- we are using the contestationPeriod as tx close grace time. See ADR21 for context
+      closeGraceTime = SlotNo $ fromIntegral cpNatural
```
Ok this is also puzzling a bit. Previously we had `closeGraceTime` expressed in slots and hardcoded to 100 slots. That was fed into the `adjustPointInTime` for close/contest tx. Right now we have CP as number of seconds and we want to use that instead of `closeGraceTime`. I don't see a way of converting number of seconds to slots immediately. I tried adding CP to current slot and use that for tx validity but the tests are not happy that way. I'll try adding CP value to currentTime and then convert it to slots to see if that helps.

v0d1ch replied:
Ok, that is just wrong. What we need here is to actually _divide_ cp seconds to the slot length for the underlying network in order to get what these seconds mean in slots but that doesn't seem possible with the current code. Maybe extend the `TimeHandle` ?

v0d1ch replied:
Another simpler option - subtract the calculated end slot and current slot to get the slot difference.

v0d1ch replied:
Ok this approach worked. I'll stop rambling.

v0d1ch committed:
Change the way we calculate tx upper bound

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -95,22 +96,22 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
   refuelIfNeeded tracer node Alice 100_000_000
   -- Start hydra-node on chain tip
   tip <- queryTip networkId nodeSocket
+  let contestationPeriod = UnsafeContestationPeriod 2
```
This is WAAY too short for this scenario, which also is used as our smoke test scenario on public testnets! That's also why there were `600` seconds timeouts.

It's annoying, but I think we need to parameterize the scenario by the block time (0.1 seconds on our current devnet and 20 seconds on testnets) to keep fast local tests, but still allow the scenarios to be run on public testnets as well.

This can be achieved by adding a `blockTime` to `RunningNode` or so.

v0d1ch replied:
I've reverted  back to the original values and tests pass. Do you think `blockTime` in `RunningNode` should be part of this PR or a follow up?

v0d1ch replied:
Ok, we can tackle the network aspect of tests in a separate issue.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -95,11 +96,15 @@ mkTimeHandle now systemStart eraHistory = do
+  currentTipSlot <-
+    case tip of
+      ChainPointAtGenesis -> throwIO $ QueryNodeException "Network is not synced yet!"
+      ChainPoint slotNo _ -> pure slotNo
```
Why are we failing here? If our node is at genesis, it's not a problem per se.

I think we should be okay by not doing this check and also not introducing this exception (btw the `Text` in the type was smell, which led me here)

v0d1ch replied:
If we are at genesis I should set the slot number to 1? 0? Both values produce green tests probably because we are passed genesis.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,24 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+  -- Get the difference between tx upper bound and current slot to determine
+  -- for how many slots this tx is valid.
+  -- NB: Used in 'adjustPointInTime' to calculate tx upper bound.
+  let closeGraceTime = txEndSlot - currentSlot
```
Do these calculations only when necessary.

Should: move them to the Close case block.

ch1bo replied:
Also: We would not need to do `adjustPointInTime` because we would already have the upper bound slot+time here: `addUTCTime (fromIntegral cpNatural) currentTime` is the `UTCTime` and `txEndSlot` is the slot.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,24 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+  pointInTime@(currentSlot, currentTime) <- throwLeft currentPointInTime
+  let (UnsafeContestationPeriod cpNatural) = contestationPeriod ctx
+  -- calculate tx upper bound by using contestation period and current time. See ADR21 for context
+  txEndSlot <- throwLeft $ slotFromUTCTime $ addUTCTime (fromIntegral cpNatural) currentTime
```
Shit.. this will become problematic if contestation periods are (rightfully) high. When the CP is > the horizon (`3k` slots iirc) this will fail.

:thinking:

ch1bo replied:
Maybe we can extend the `TimeHandle` to give us the known "horizon". Then we can cap the upper bound by this time/slot, i.e. `min(horizon, deadline)` with ``deadline = period `addUTCTime` currenTime``

v0d1ch replied:
We ended up not extending the `TimeHandle` and just adding `min (maxGraceTime, contestationPeriod)` where `maxGraceTime` is reasonable value for devnet.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -230,21 +231,24 @@ chainSyncHandler tracer callback getTimeHandle ctx =
-  pointInTime <- throwLeft currentPointInTime
+prepareTxToPost timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
+  pointInTime@(currentSlot, currentTime) <- throwLeft currentPointInTime
+  let (UnsafeContestationPeriod cpNatural) = contestationPeriod ctx
```
We should provide a `ContestationPeriod -> NominalDiffTime` conversion and then it would not look as unsafe as this.

v0d1ch replied:
It even exists already https://github.com/input-output-hk/hydra/blob/11c015bbe3a28def0935e247eb0d54af17f47948/hydra-node/src/Hydra/ContestationPeriod.hs#L53

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -472,7 +481,13 @@ observeInit ::
+    observeInitTx
+      networkId
+      (allVerificationKeys ctx)
+      (Hydra.Chain.Direct.State.contestationPeriod ctx)
```
Could: Access by record field in line 505 to make this less awkward.

v0d1ch replied:
it clashes with `InitObservation` `contestationPeriod`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -834,18 +850,21 @@ genCloseTx numParties = do
-  pure (cctx, stOpen, close cctx stOpen snapshot pointInTime, snapshot)
+  let cp = ctxContestationPeriod ctx
+  (startSlot, pointInTime) <- genValidityBoundsFromContestationPeriod cp
+  -- The difference between startSlot and pointInTime needs to be <= contestationPeriod
```
This is already documented in `close` haddock

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -42,33 +44,32 @@ data TimeHandle = TimeHandle
 -- | Generate consistent values for 'SystemStart' and 'EraHistory' which has
 -- a horizon at the returned SlotNo as well as some UTCTime before that
-genTimeParams :: Gen (SystemStart, EraHistory CardanoMode, SlotNo, UTCTime)
+genTimeParams :: Gen (SystemStart, EraHistory CardanoMode, SlotNo, SlotNo)
```
Should: Create a small record type to make these two `SlotNo` distinguishable.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -42,33 +44,32 @@ data TimeHandle = TimeHandle
+    (systemStart, eraHistory, _, currentSlotNo) <- genTimeParams
+    pure $ mkTimeHandle currentSlotNo systemStart eraHistory
 
 -- | Construct a time handle using current time and given chain parameters. See
```
```suggestion
-- | Construct a time handle using current slot and given chain parameters. See
```

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -135,3 +142,10 @@ genConfirmedSnapshot minSn utxo sks
     let signatures = aggregate $ fmap (`sign` snapshot) sks
     pure $ ConfirmedSnapshot{snapshot, signatures}
+
+fromChainSnapshot :: HS.SnapshotNumber -> SnapshotNumber
```
Could: Import qualified as `OnChain` to make this even more readable.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -170,11 +178,17 @@ genCloseMutation (tx, _utxo) =
+    , -- try to change a tx so that lower bound is higher than the upper bound
+      SomeMutation MutateValidityInterval . ChangeValidityInterval <$> do
+        lb <- arbitrary
+        ub <- arbitrary `suchThat` (< lb)
```
More robust would be to `(lb -) <$> choose (0, lb)`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -200,12 +214,12 @@ genCloseMutation (tx, _utxo) =
-  mutateClosedContestationDeadline :: Gen (TxOut CtxTx)
-  mutateClosedContestationDeadline = do
+  -- In case contestation period param is 'Nothing' we will generate arbitrary value
+  mutateClosedContestationDeadline :: Maybe Integer -> Gen (TxOut CtxTx)
```
Instead of passing `Maybe` (and conditionally switching logic inside the funciton) I would recommend to pass the `Integer` and generate it outside.

ch1bo replied:
Still relevant (also the comment above). But this is a "Could".

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TimeHandleSpec.hs:
```diff
@@ -21,10 +21,10 @@ spec = do
 
   it "should convert slot within latest/current era" $ do
-    let now = posixSecondsToUTCTime 13.4
+    let currentSlotNo = SlotNo 10
```
The equivalent would be `13`, but this is fine as well .. I hope (can't recall this unit test in detail)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -194,6 +196,21 @@ spec =
 
+      prop "Ignore InitTx with wrong contestation period" $
+        withMaxSuccess 60 $ \txIn cperiod (party :| parties) cardanoKeys -> do
+          i <- arbitrary `suchThat` (> 0)
```
Could: use `Positive` (and avoid `suchtThat` per our new coding guideline), i.e. `i <- getPositive <$> arbitrary`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -194,6 +196,21 @@ spec =
+          let params = HeadParameters cperiod (party : parties)
+              (UnsafeContestationPeriod cp) = cperiod
+              -- construct different/wrong CP
+              wrongCPeriod = UnsafeContestationPeriod $ cp + wordToNatural i
```
Should: just generate a `i :: Natural`?

v0d1ch replied:
We have overlapping instances for Arbitrary Natural in scope here and it is not so easy to resolve this error so I didn't want to waste any time.

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +557,40 @@ hydraNodeCommand =
       (long "script-info" <> help "Dump script info as JSON")
 
+defaultContestationPeriod :: ContestationPeriod
+defaultContestationPeriod = UnsafeContestationPeriod 3
```
As I said before: 3 is a bad default! It's too low for most networks. Use `60`

v0d1ch committed:
Constrain arbitrary instance for cp in MBT

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -550,6 +557,40 @@ hydraNodeCommand =
+    case eparsed of
+      Right 0 -> Left "contestation-period argument needs to be higher than zero"
+      Right a -> Right $ UnsafeContestationPeriod a
+      Left _ -> Left errorMsg
```
How about: `parseNatural <|> parseNominalDiffTime` with `parseNatural = UnsafeContestationPeriod <$> auto` and

```
parseNominalDiffTime = auto >>= \diffTime -> do
  let s = nominalDiffTimeToSeconds diffTime
  if s <= 0
  then fail "negative contestation period"
  else UnsafeContestationPeriod $ fromIntegral s
```
?

ch1bo replied:
(I usually think in combining cases with backtracing, rather scrutinizing for which case it is -> looking for the `'s'`)

v0d1ch replied:
yes, looks a lot nicer

ch1bo commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -105,9 +107,9 @@ spec = parallel $ do
       let events =
             [ NetworkEvent{ttl = defaultTTL, message = Connected{nodeId = NodeId "NodeId1"}}
             , NetworkEvent{ttl = defaultTTL, message = Connected{nodeId = NodeId "NodeId2"}}
-            , ClientEvent $ Init cperiod
```
Is `cperiod` still used?

v0d1ch replied:
This must be an old diff since I removed `cperiod` from `Init` already.

ch1bo commented on hydra-tui/test/Hydra/TUISpec.hs:
```diff
@@ -38,15 +38,19 @@ import Hydra.Cluster.Fixture (
 import HydraNode (EndToEndLog, HydraClient (HydraClient, hydraNodeId), withHydraNode)
 import System.Posix (OpenMode (WriteOnly), closeFd, defaultFileFlags, openFd)
 
+-- TODO(SN): hardcoded contestation period used by the tui
```
I think this is not a TODO anymore -> no problem to use hard-coded values in tests.

ch1bo reviewed (changes requested):
We need to find a way to not convert slots too far in the future.

Maybe keeping a grace time which serves as a `max(cp, graceTime)` to not try to use an upper bound too far in the future?

v0d1ch committed:
Bring back the old values in Scenarios module

v0d1ch committed:
Do not throw if we are at genesis

v0d1ch committed:
Do not use adjustPointInTime but calculate tx end slot from current time

v0d1ch committed:
Use getPositive in TxSpec

v0d1ch committed:
Use 60 seconds as default contestation period

v0d1ch committed:
Use suggestion for CP parsing

v0d1ch committed:
Fix cp parsing and remove redundant todo item

v0d1ch committed:
Add a comment regarding start slot construction on devnet

v0d1ch committed:
Remove adjustPointInTime

v0d1ch committed:
Small refactor

v0d1ch committed:
Introduce failing test case

v0d1ch committed:
Add HasCallstack to mkTimeHandle

ch1bo committed:
Limit the upper bound time (& slot) to be max 60s into the future

This is required if we have long contestation periods which would exceed
the time horizon in which we can convert time <-> slot.

60 seconds seems reasonable for all networks we know right now.

Lower contestation periods will still use that as the delay and a user
will need to pick the right --contestation-period for a network anyways.

v0d1ch committed:
Update adr21 document

v0d1ch committed:
Increase maxGraceTime to 200

v0d1ch committed:
Add new error constructor when we can't find the marked utxo

ch1bo committed:
Add api spec to new error case and add descriptions

v0d1ch committed:
Sort in unoptimal manner

v0d1ch committed:
Update changelog a bit

v0d1ch committed:
Fix CI smoke tests

v0d1ch committed:
Fix duplicated function name

v0d1ch committed:
Modify changelog and api.yaml

v0d1ch committed:
CheckPoint

v0d1ch committed:
Fix golden file

ch1bo commented on CHANGELOG.md:
```diff
@@ -34,6 +34,19 @@ changes.
+  + There is a new hydra-node flag `--contestation-period` expressed in seconds
+    to control the close tx validity bounds as well as determine the
+    contestation deadline.
+  + The deadline is always `2 * --contestation-period` after the Close tx.
```
this is not true anymore with `min(maxGraceTime, contestationPeriod)`. Instead it may be "up to 2*cp"

ch1bo commented on CHANGELOG.md:
```diff
@@ -34,6 +34,19 @@ changes.
+- **BREAKING** Change the way tx validity and contestation deadline is constructed for close transactions:
+  + There is a new hydra-node flag `--contestation-period` expressed in seconds
+    to control the close tx validity bounds as well as determine the
+    contestation deadline.
```
We should add an example: `e.g. with `--contestation-period 30s`, the node will close the head `30s` after submitting the close transaction and other parties will have another `30s` to contest.

And/or add a section in the user manual (where we describe also the fuel stuff e.g) about this.

ch1bo commented on docs/adr/2022-12-05_021-Bounded-transaction-validity-on-Hydra-protocol-transactions.md:
```diff
@@ -0,0 +1,102 @@
+  appropriate value.
+
+- The upper tx validity of close transaction should be minimum between
+  `maxGraceTime` and `contestationPeriod`. Users need to know this.
```
Is this now context, decision or consequence? It's mentioned everywhere :thinking:

ch1bo replied:
I'd say it's really a consequence of the fact that we cannot convert arbitrary point in times to slots (which is required by the ledger when constructing transactions).

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -58,7 +60,7 @@ import Test.QuickCheck (
   suchThat,
   vectorOf,
-  withMaxSuccess,
+  withMaxSuccess, getPositive
```
No formatter in use?

v0d1ch replied:
seems so. I re-saved the file and it worked.

ch1bo commented on nix/hydra/shell.nix:
```diff
@@ -135,6 +135,7 @@ let
       cardano-node.cardano-node
       cardano-node.cardano-cli
       hsPkgs.hydra-node.components.exes.hydra-node
+      hsPkgs.hydra-cluster.components.exes.hydra-cluster
```
Not needed.. right? The `exes` shell includes it already. Or where you confused which one to pick?

v0d1ch replied:
Leftover from smoke tests CI debug...ooops

ch1bo commented on hydra-node/src/Hydra/ContestationPeriod.hs:
```diff
@@ -20,13 +20,11 @@ instance Arbitrary ContestationPeriod where
-        , pure oneWeek
-        , pure oneMonth
-        , pure oneYear
+        , elements [oneDay, oneWeek, oneMonth, oneYear]
```
FWIW this is not strictly the same as the old code! It will change the frequency of these values. I think it's acceptable, but I wonder why you did change it.

v0d1ch replied:
It was part of previous _experiments_. I am reverting it to the old instance we had.

ch1bo reviewed (approved):
I think this is good to go! Minor comments and some documentation would be perfect, but also fine if you want to get it merged now.

v0d1ch committed:
PR review final comments
